File: Arrays&Hashing/group-anagrams.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>

using namespace std;
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
		map<string, vector<string>> m;
		for (auto s : strs) {
			string sorted = s;
			sort(sorted.begin(), sorted.end());
			m[sorted].push_back(s);
		}
		vector<vector<string>> res;
		for (auto p : m) {
			res.push_back(p.second);
		}
		return res;
    }
};

File: Arrays&Hashing/group-anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
// this code worrks but is slow
let groupIndex = (str, ret) => {
    for (let i = 0; i < ret.length; i++) {
        if (ret[i][0].split``.sort().join`` == str)
            return i;
    }
    return -1;
}


var groupAnagrams = function(strs) {
    let ret = [];
    for (i in strs) {
        let tmp = strs[i].split``.sort().join``;
        if (groupIndex(tmp, ret) == -1) // is new
            ret.push([strs[i]]);
        else // exists somewhere
            ret[groupIndex(tmp, ret)].push(strs[i]);
    }
    return ret;
};
